core/src/main/java/com/linecorp/armeria/common/util/Unwrappable.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common.util;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Optional;
+
+/**
+ * Provides a way to unwrap an object in decorator pattern, as an alternative to down-casting.
```

Comment:
I don't think it's an alternative, so maybe `similar to down-casting in an inheritance pattern`.

Reply:
Updated.